### PR TITLE
Llama3 pre-context parallel dataloader changes

### PR DESCRIPTION
**Pull request by pstjohn**
Makes a number of updates in preparation for llama3 context-parallel training. It's still not currently working, need to further update the model to handle the `cu_seq_lens_q_padded` kwargs and would like to add a single-GPU CP test that uses BSHD inputs to at least exercise this code in CI.

This PR:
* Only materializes the dataloader on the cp_rank=0, and returns None on other ranks.
* Uses the scatter operation in the dataloader to synchronize `StopIteration` exceptions
* Adds tests for the CP dataloader on 1 and 2-gpu machines
* moves llama3 to use DLCM data as the sanity dataset, turns off some genome collation options by default. This is larger than the dummy sequences currently used in training, and will make sure we can fill a few batches in CP testing. We may want to revert this eventually once we're done bringing up llama3; since it does trigger the tokenizer download during testing.
* removes `lazy tokenization` from llama3, this wont work. See https://nvidia.slack.com/archives/C074Z808N05/p1767818883160949
* starts adding CP files for llama3

Closes BIO-11

### CODE REVIEW

File: bionemo-recipes/recipes/llama3_native_te/dataset.py [hunk 1]
```diff
@@ -116,11 +116,10 @@ def create_bshd_dataloader(
...
     seed: int = 42,
     buffer_size: int = 500_000,
-    use_lazy_tokenization: bool = True,
     use_stateful_dataloader: bool = False,
     text_column: str = "text",
     uppercase_labels: bool = False,
-    mask_degenerate_bases: bool = True,
+    mask_degenerate_bases: bool = False,
 ):
     """Create a BSHD dataloader for genomic sequences using CLM (causal language modeling).
```
Comment (savitha-eng) [inline bionemo-recipes/recipes/llama3_native_te/dataset.py]: remove genomic here

Comment (savitha-eng) [inline bionemo-recipes/recipes/llama3_native_te/dataset.py]: nit: maybe get rid of all references to genomic sequences in comments, since we are now defaulting to the dclm text dataset

File: bionemo-recipes/recipes/llama3_native_te/tests/test_dataset.py [hunk 2]
```diff
@@ -123,9 +134,8 @@ def test_dataloader_returns_expected_batch(tokenizer_path, tmp_path):
         load_dataset_kwargs=load_dataset_kwargs,
         micro_batch_size=1,  # Just one sample per batch
         num_workers=0,
-        max_seq_length=10,  # Large enough for 5bp sequence
+        max_seq_length=7,  # Large enough for 5bp sequence
```
Comment (savitha-eng) [inline bionemo-recipes/recipes/llama3_native_te/tests/test_dataset.py:137]: Why was max_seq_length changed from 10 to 7 ? Is this related to CP padding requirements?

File: bionemo-recipes/recipes/llama3_native_te/checkpoint.py [hunk 3]
```diff
@@ -328,13 +328,15 @@ def save_checkpoint_fsdp2(
...
         logger.info(f"Saved FSDP2 dataloader to {ckpt_path}")
 
-    # If we're using asynchronous checkpointing, make sure we only have one checkpoint future at a time.
-    if async_save and "fsdp2" in _ckpt_futures and _ckpt_futures["fsdp2"] is not None:
-        _ckpt_futures["fsdp2"].result()
-
     state_dict = {"app": AppState(model=model, optimizer=optimizer, scheduler=scheduler, step=step, epoch=epoch)}
-    ckpt_save_func = dcp_async_save if async_save else dcp_save
-    _ckpt_futures["fsdp2"] = ckpt_save_func(state_dict, checkpoint_id=checkpoint_path, process_group=process_group)
+    if async_save:
```
Comment (savitha-eng) [inline bionemo-recipes/recipes/llama3_native_te/checkpoint.py:332]: is this related specifically to CP or just a general bug fix?

File: bionemo-recipes/recipes/llama3_native_te/tests/test_dataset.py (hunk 2 above)
Comment (pstjohn) [reply bionemo-recipes/recipes/llama3_native_te/tests/test_dataset.py:137]: i think it was just so the test logic didn't have to deal with padding tokens

File: bionemo-recipes/recipes/llama3_native_te/checkpoint.py (hunk 3 above)
Comment (pstjohn) [reply bionemo-recipes/recipes/llama3_native_te/checkpoint.py:332]: general bug fix, i was seeing some failures related to not properly cleaning up this async futures dictionary in our test suite

File: bionemo-recipes/recipes/llama3_native_te/dataset.py (hunk 1 above)
Comment (pstjohn) [reply bionemo-recipes/recipes/llama3_native_te/dataset.py]: done!
